sqlx-sqlite/src/index.rs

```diff
@@ -0,0 +1,599 @@
+    ///
+    /// The filter is pushed down to the index by converting it to a set of SQL expressions
+    /// that can be evaluated by the index.
+    ///
+    /// The return value is a list of `(file_name, FileScanPlan)` tuples where `FileScanPlan` contains
+    /// the file metadata and the row groups that need to be scanned.
+    pub async fn get_files(
+        &self,
+        filter: Arc<dyn PhysicalExpr>,
+        schema: SchemaRef,
+    ) -> Result<Vec<(String, FileScanPlan)>> {
+        // Convert the predicate to a pruning predicate
+        // This transforms e.g. `a = 5` to `a_min <= 5 AND a_max >= 5`
+        let pruning = PruningPredicate::try_new(filter, schema.clone())?;
+        let predicate = pruning.predicate_expr().clone();
+        // Replace any `{col}_row_count` with `row_count` as we don't store per-column row counts
+        // (they're the same for all columns in a row group)
+        let predicate = predicate.transform(|expr| {
+            if let Some(column) = expr.as_any().downcast_ref::<phys_expr::Column>() {
+                if column.name().ends_with("_row_count") {
+                    let column = phys_expr::Column::new(column.name().trim_end_matches("_row_count"), column.index());
+                    return Ok(Transformed::yes(Arc::new(column)));
+                }
+            }
+            Ok(Transformed::no(expr))
+        }).data()?;
+        // Convert a DataFusion PhysicalExpr to a SeaQuery SimpleExpr
+        let predicate = physical_expr_to_sea_query(&predicate);
+
+        let stats_query = Query::select()
+            .from(Alias::new("row_group_statistics"))
+            .columns(vec![
+                Alias::new("file_id"),
+                Alias::new("row_group"),
+            ])
+            .and_where(predicate).to_owned();
+        
+        let cte = CommonTableExpression::new()
+            .query(stats_query)
+            .table_name(Alias::new("row_groups")).to_owned();
+
+        let files_query = Query::select()
+            .from(Alias::new("file_statistics"))
+            .columns(vec![
+                Alias::new("file_name"),
+                Alias::new("file_size_bytes"),
+                Alias::new("row_group_count"),
+            ])
+            .inner_join(
+                Alias::new("row_groups"), 
+                SeaQExpr::col((Alias::new("file_statistics"), Alias::new("file_id"))).equals((Alias::new("row_groups"), Alias::new("file_id"))),
+            )
+            .column(Alias::new("row_group"))
+            .distinct()
+            .to_owned();
+
+        let query = files_query.with(WithClause::new().cte(cte).to_owned());
+
+        let (sql, values) = query.build_sqlx(SqliteQueryBuilder);
+
+        let row_groups: Vec<(String, i64, i64, i64)> = sqlx::query_as_with(&sql, values)
+            .fetch_all(&self.pool)
+            .await
+            .unwrap(); // TODO: handle error, possibly failing gracefully by scanning all files?
```

Comment:
For sake of readability of the example it might be nice to use a more clear type than the tuple here for each row, by making a struct and using `#[derive(sqlx::FromRow)]`.

Reply:
a6c1d69